src/metaprob/syntax.clj

```diff
@@ -124,11 +126,45 @@
                       '~(seq (map str names))
                       ~names)))
 
+(defn- inline-list
+  "Returns Metaprob code for building a trace with the same structure as the
+  provided list."
+  [xs]
+  (if-let [s (seq xs)]
+    `(trace-set-subtrace (trace-set-value (empty-trace)
+                                          ~(first s))
+                         ~state/rest-marker
+                         ~(build-list (rest s)))
+    `(trace)))
+
+(defn- inline-map
+  "Returns Metaprob code for building a trace with the same structure as the
+  provided map."
+  [x]
+  (reduce-kv (fn [acc k v]
+               `(trace-set-subtrace ~acc ~k ~v))
+             `(trace)
+             x))
+
+(defn- inline-collection
+  ""
+  [x]
+  (cond (map? x)
+        (inline-map x)
+
+        (and (seq? x)
+             (= 'list (first x)))
+        (inline-list (rest x))
+
+        :else x))
+
 (defmacro gen
   "like fn, but for metaprob procedures"
   {:style/indent 1}
   [params & body]
-  `(named-generator nil ~params ~@body))
+  `(named-generator nil
+                    ~params
+                    ~@(walk/postwalk inline-collection body)))
```

Comment:
If the point is to add productions to the metaprob-in-clojure syntax, why not just modify the main dispatch?

I've never heard of `walk/postwalk` and it sounds like it would have all sorts of problems (wrong order of growth, nasty error messages, etc)

Why are you using `defn-` instead of `defn` ?

Reply:
> why not just modify the main dispatch

I was attempting to adhere to my understanding of Alex's proposal in [this comment](https://github.com/probcomp/metaprob-clojure/pull/76#issuecomment-427691456). Just modifying the main dispatch seems reasonable to me, and is what I was asking about [here](https://github.com/probcomp/metaprob-clojure/pull/76#issuecomment-427723348). Thanks for clarifying!

Just so I can learn, could you say more about what problems you imagine `clojure.walk/postwalk` might have? It executes a postwalk traversal of the source, so I would expect that the order of growth would be linear relative to the size of the source tree. 

As for `defn-`, I thought that we would want these functions to be private.

Reply:
It's hard for me to see that use of a code walker wouldn't interfere in some foul way with the ability to run gen bodies using the metaprob interpreter. Maybe not.

The code walker gets run every time a gen expression is macroexpanded, so the expanded code would get walked a number of times proportional to the gen nesting depth. Having a code walker touch a given piece of code (and thereby assuming that it's idempotent) seems sort of nutty to me.

In my experience code walkers are almost always the wrong thing. I haven't looked to see what this particular one does, but if there is a better way to do this or any other task, I'd stay away from code walkers.

Reply:
that was incoherent... try "Having a code walker touch a piece of code multiple times"